From b0d4b344222613e4ee35615a370a216a5e2ee7df Mon Sep 17 00:00:00 2001
From: Sujeet Baranwal <sbaranwal@marvell.com>
Date: Mon, 6 May 2019 16:53:06 -0700
Subject: [PATCH 0404/1921] mmc: octeontx2: emmc operation limited to 100 MHz

Because of errata on otx2, the emmc clock is limited
to 100MHz.

Change-Id: Ide1c57fec1c119b3a952a50a056b5236a4b22453
Signed-off-by: Sujeet Baranwal <sbaranwal@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/8532
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/26899
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/mmc/host/cavium.c | 33 ++++++++++++++++++++++++++++-----
 drivers/mmc/host/cavium.h |  5 +++++
 2 files changed, 33 insertions(+), 5 deletions(-)

diff --git a/drivers/mmc/host/cavium.c b/drivers/mmc/host/cavium.c
index aa586e6644ba..04b46a595697 100644
--- a/drivers/mmc/host/cavium.c
+++ b/drivers/mmc/host/cavium.c
@@ -826,11 +826,28 @@ static void cvm_mmc_request(struct mmc_host *mmc, struct mmc_request *mrq)
 	writeq(emm_cmd, host->base + MIO_EMM_CMD(host));
 }
 
+static u32 max_supported_frequency(struct cvm_mmc_host *host)
+{
+	/* Default maximum freqeuncey is 52000000 for chip prior to 9X */
+	u32 max_frequency = MHZ_52;
+
+	if (is_mmc_otx2(host)) {
+		/* Default max frequency is 200MHz for 9X chips */
+		max_frequency = MHZ_200;
+
+		/* Erratum is only applicable pass A0 */
+		if (is_mmc_otx2_A0(host))
+			max_frequency = MHZ_100;
+	}
+	return max_frequency;
+}
+
 static void cvm_mmc_set_ios(struct mmc_host *mmc, struct mmc_ios *ios)
 {
 	struct cvm_mmc_slot *slot = mmc_priv(mmc);
 	struct cvm_mmc_host *host = slot->host;
 	int clk_period = 0, power_class = 10, bus_width = 0;
+	u32 max_f;
 	u64 clock, emm_switch;
 
 	if (ios->power_mode == MMC_POWER_OFF) {
@@ -875,8 +892,11 @@ static void cvm_mmc_set_ios(struct mmc_host *mmc, struct mmc_ios *ios)
 
 	/* Change the clock frequency. */
 	clock = ios->clock;
-	if (clock > 52000000)
-		clock = 52000000;
+	max_f = max_supported_frequency(host);
+
+	if (clock > max_f)
+		clock = max_f;
+
 	slot->clock = clock;
 
 	if (clock)
@@ -959,6 +979,7 @@ static int cvm_mmc_of_parse(struct device *dev, struct cvm_mmc_slot *slot)
 	u32 id, cmd_skew = 0, dat_skew = 0, bus_width = 0;
 	struct device_node *node = dev->of_node;
 	struct mmc_host *mmc = slot->mmc;
+	u32 max_frequency;
 	int ret;
 
 	ret = of_property_read_u32(node, "reg", &id);
@@ -1002,12 +1023,14 @@ static int cvm_mmc_of_parse(struct device *dev, struct cvm_mmc_slot *slot)
 			mmc->caps |= MMC_CAP_4_BIT_DATA;
 	}
 
+	max_frequency = max_supported_frequency(slot->host);
+
 	/* Set maximum and minimum frequency */
 	if (!mmc->f_max)
 		of_property_read_u32(node, "spi-max-frequency", &mmc->f_max);
-	if (!mmc->f_max || mmc->f_max > 52000000)
-		mmc->f_max = 52000000;
-	mmc->f_min = 400000;
+	if (!mmc->f_max || mmc->f_max > max_frequency)
+		mmc->f_max = max_frequency;
+	mmc->f_min = KHZ_400;
 
 	/* Sampling register settings, period in picoseconds */
 	of_property_read_u32(node, "cavium,cmd-clk-skew", &cmd_skew);
diff --git a/drivers/mmc/host/cavium.h b/drivers/mmc/host/cavium.h
index 539da17b27ce..172edad3a65f 100644
--- a/drivers/mmc/host/cavium.h
+++ b/drivers/mmc/host/cavium.h
@@ -28,6 +28,11 @@
 #define PCI_SUBSYS_DEVID_8XXX	0xA
 #define PCI_SUBSYS_DEVID_9XXX	0xB
 
+#define KHZ_400 (400000)
+#define MHZ_52  (52000000)
+#define MHZ_100 (100000000)
+#define MHZ_200 (200000000)
+
 /* DMA register addresses */
 #define MIO_EMM_DMA_FIFO_CFG(x)	(0x00 + x->reg_off_dma)
 #define MIO_EMM_DMA_FIFO_ADR(x)	(0x10 + x->reg_off_dma)
-- 
2.31.1

